pkg/controller/sparkapplication/batchscheduler/volcano/volcano_scheduler.go

```diff
@@ -0,0 +1,162 @@
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupForClientAPP(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We only care about the executor pods in client mode
+	if _, ok := app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		if err := v.syncPodGroup(app, *app.Spec.Executor.Instances); err == nil {
+			app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		}
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroupForClusterAPP(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	//We need both mark Driver and Executor when submitting
+	//NOTE: Although we only have one pod for Spark Driver, we still manage it into PodGroup,since it can
+	//utilize other advanced features in volcano, for instance, namespace resource fairness.
+	if _, ok := app.Spec.Driver.Annotations[v1alpha2.GroupNameAnnotationKey]; !ok {
+		if err := v.syncPodGroup(app, 1); err != nil {
+			app.Spec.Executor.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+			app.Spec.Driver.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		}
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) OnSparkDriverPodScheduled(app *v1beta1.SparkApplication) (*v1beta1.SparkApplication, error) {
+	// Update PodGroup MinMember if required
+	if app.Spec.Mode == v1beta1.ClusterMode {
+		return app, v.syncPodGroup(app, 1+*(app.Spec.Executor.Instances))
+	}
+	return app, nil
+}
+
+func (v *VolcanoBatchScheduler) getAppPodGroupName(app *v1beta1.SparkApplication) string {
+	return fmt.Sprintf("sparkapplication-%s-podgroup", app.Name)
+}
+
+func (v *VolcanoBatchScheduler) syncPodGroup(app *v1beta1.SparkApplication, size int32) error {
+	var err error
+	podGroupName := v.getAppPodGroupName(app)
+	if pg, err := v.volcanoClient.SchedulingV1alpha2().PodGroups(app.Namespace).Get(podGroupName, v1.GetOptions{}); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		podGroup := v1alpha2.PodGroup{
+			ObjectMeta: v1.ObjectMeta{
+				Namespace: app.Namespace,
+				Name:      podGroupName,
+				OwnerReferences: []v1.OwnerReference{
+					*v1.NewControllerRef(app, v1beta1.SchemeGroupVersion.WithKind("SparkApplication")),
+				},
+			},
+			Spec: v1alpha2.PodGroupSpec{
+				MinMember: size,
+			},
+		}
+
+		_, err = v.volcanoClient.SchedulingV1alpha2().PodGroups(app.Namespace).Create(&podGroup)
+	} else {
+		if pg.Spec.MinMember != size {
+			pg.Spec.MinMember = size
+			_, err = v.volcanoClient.SchedulingV1alpha2().PodGroups(app.Namespace).Update(pg)
+		}
+	}
+	if err != nil {
+		glog.Errorf(
+			"Unable to sync PodGroup with error: %s. Abandon schedule pods via volcano.", err)
+	}
+	return err
+}
+
+func New(config *rest.Config) schedulerinterface.BatchScheduler {
+
+	vkClient, err := volcanoclient.NewForConfig(config)
+	if err != nil {
+		glog.Errorf("Unable to initialize volcano client with error %v", err)
```

Comment:
We typically use lower case for error messages and error messages like `failed to ...: %v`. In this case `failed to initialize the valcano client: %v`.